src/Microsoft.Azure.SignalR.Common/ClientInvocation/ClientResultsManager.cs

```diff
@@ -0,0 +1,323 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if NET7_0_OR_GREATER
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.SignalR.Protocol;
+using System.Buffers;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class ClientResultsManager: IClientResultsManager, IInvocationBinder
+    {
+        private readonly ConcurrentDictionary<string, (Type Type, string ConnectionId, object Tcs, Action<object, CompletionMessage> Complete)> _pendingInvocations = new();
+        private ulong _lastInvocationId = new();
+
+        private readonly ConcurrentDictionary<string, (Type Type, string ConnectionId, string CallerServerId, object Tcs, Action<object, CompletionMessage> Complete)> _routedInvocations = new();
+
+        private readonly ConcurrentDictionary<string, ServiceMappingMessage> _serviceMappingMessages = new();
+        private readonly IHubProtocolResolver _hubProtocolResolver;
+
+        public ClientResultsManager(IHubProtocolResolver hubProtocolResolver)
+        {
+            _hubProtocolResolver = hubProtocolResolver;
+        }
+
+        public ulong GetNewInvocation()
+        {
+            return Interlocked.Increment(ref _lastInvocationId);
+        }
+
+        public void AddServiceMappingMessage(string instanceId, ServiceMappingMessage serviceMappingMessage)
+        {
+            _serviceMappingMessages.TryAdd(instanceId, serviceMappingMessage);
+        }
+
+        public void RemoveServiceMappingMessageWithOfflinePing(string instanceId)
+        {
+            if (_serviceMappingMessages.TryRemove(instanceId, out var serviceMappingMessage))
+            {
+                if (_pendingInvocations.TryRemove(serviceMappingMessage.InvocationId, out var item))
+                {
+                    var message = new CompletionMessage(serviceMappingMessage.InvocationId, "mapping service offline", null, true);
+                    item.Complete(item.Tcs, message);
+                }
+            }
+        }
+
+        public Task<T> AddRoutedInvocation<T>(string connectionId, string invocationId, string callerServerId, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSourceWithCancellation<T>(this, connectionId, invocationId, cancellationToken);
+            var result = _routedInvocations.TryAdd(invocationId, (typeof(T), connectionId, callerServerId, tcs, static (state, completionMessage) =>
+            {
+                var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+                if (completionMessage.HasResult)
+                {
+                    tcs.SetResult((T)completionMessage.Result);
```

Comment:
For routed invocation, looks we fake the type as `string` always when call `AddRoutedInvocation`. It's not correct here to convert type...

Reply:
Fixed in #1684